### PR TITLE
Add functions to assist in immutability of Event objects

### PR DESCRIPTION
**Pull request by turt2live**
**For https://github.com/matrix-org/matrix-react-sdk/pull/6196**

### CODE REVIEW

File: spec/unit/utils.spec.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import * as utils from "../../src/utils";
 import {
     alphabetPad,
     averageBetweenStrings,
-    baseToString,
+    baseToString, deepSortedObjectEntries,
```
Comment (jryans) [inline spec/unit/utils.spec.ts:5]: Maybe on separate lines to match the rest here...? Or else smush the others into something more compact...?

File: src/models/event.js [hunk 2]
```diff
@@ -1143,6 +1144,41 @@ utils.extend(MatrixEvent.prototype, {
     getTxnId() {
         return this._txnId;
     },
+
+    /**
+     * Get a copy/snapshot of this event. The returned copy will be loosely linked
+     * back to this instance, though will have "frozen" event information. Other
+     * properties may mutate depending on the state of this instance at the time
```
Comment (jryans) [inline src/models/event.js:1151]: Hmm, what properties are you thinking of...?

File: src/models/event.js [hunk 3]
```diff
@@ -1143,6 +1144,41 @@ utils.extend(MatrixEvent.prototype, {
...
+     * of snapshotting.
+     *
+     * This is meant to be used to snapshot the event details themselves, not the
+     * features (such as sender) surrounding the event.
+     * @returns {MatrixEvent} A snapshot of this event.
+     */
+    getSnapshotCopy() {
```
Comment (jryans) [inline src/models/event.js:1158]: Maybe `toSnapshot`, similar to `toJSON`? `SnapshotCopy` seems redundant to me...

```suggestion
    toSnapshot() {
```

File: src/models/event.js [hunk 4]
```diff
@@ -1143,6 +1144,41 @@ utils.extend(MatrixEvent.prototype, {
...
+                ev[p] = v;
+            }
+        }
+        return ev;
+    },
+
+    /**
+     * Determines if this event is equivalent to the given event. This only checks
+     * the event object itself, not the other properties of the event. Intended for
+     * use with getSnapshotCopy() to identify events changing.
```
Comment (jryans) [inline src/models/event.js:1171]:
```suggestion
     * use with toSnapshot() to identify events changing.
```

**Review by jryans — APPROVED** [review]
Looks good overall, but please consider tweaking the name.

File: src/models/event.js (hunk 2 above)
Comment (turt2live) [reply src/models/event.js:1151]: `sender`, mostly. Will clarify

**Comment by turt2live** [comment]
@jryans I've reworded the function docs a bit - can I get a second look?

File: src/models/event.js [hunk 5]
```diff
@@ -1143,6 +1144,42 @@ utils.extend(MatrixEvent.prototype, {
     getTxnId() {
         return this._txnId;
     },
+
+    /**
+     * Get a copy/snapshot of this event. The returned copy will be loosely linked
+     * back to this instance, though will have "frozen" event information. Other
+     * properties of this MatrixEvent instance will be copied verbatim, which can
+     * mean they are in reference to this instance despite being on the copy too.
+     * Consumers should be wary of using fields which may mutate over time.
```
Comment (jryans) [inline src/models/event.js:1153]: Hmm, the wording is still a bit strange to me, maybe include an example of the sender profile...?

**Review by jryans — APPROVED** [review]
Looks good overall, thanks! 😄

**Pull request merged**